src/connection.rs

```diff
@@ -353,3 +206,180 @@ impl From<ConnectionState> for u32 {
+
+    Ok((connection, state))
+}
+
+fn get_connection_active_path(dbus_manager: &DBusNetworkManager,
+                              connection_path: &str)
+                              -> Result<Option<String>, String> {
+    let active_paths = try!(dbus_manager.get_active_connections());
+
+    for active_path in active_paths {
+        if let Some(settings_path) = dbus_manager.get_active_connection_path(&active_path) {
+            if connection_path == settings_path {
+                return Ok(Some(active_path));
+            }
+        }
+    }
+
+    Ok(None)
+}
+
+fn wait(connection: &Connection,
+        target_state: ConnectionState,
+        timeout: u64)
+        -> Result<ConnectionState, String> {
+    if timeout == 0 {
+        return connection.get_state();
+    }
+
+    let mut total_time = 0;
+
+    loop {
+        ::std::thread::sleep(::std::time::Duration::from_secs(1));
+
+        let state = try!(connection.get_state());
+
+        total_time += 1;
+
+        if state == target_state || total_time >= timeout {
+            return Ok(state);
+        }
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::super::NetworkManager;
+
+    use super::*;
+
+    #[test]
+    fn test_connection_enable_disable() {
+        let manager = NetworkManager::new();
+
+        let connections = manager.get_connections().unwrap();
+
+        // set enviorment variable $TEST_WIFI_SSID with the wifi's SSID that you want to test
+        // e.g.  export TEST_WIFI_SSID="Resin.io Wifi"
+        let wifi_env_var = "TEST_WIFI_SSID";
+        let connection = match ::std::env::var(wifi_env_var) {
+            Ok(ssid) => {
+                connections
+                    .iter()
+                    .filter(|c| c.settings().ssid == ssid)
+                    .nth(0)
+                    .unwrap()
+                    .clone()
+            }
+            Err(e) => {
+                panic!("couldn't retrieve enviorment variable {}: {}",
+                       wifi_env_var,
+                       e)
+            }
+        };
+
+        let state = connection.get_state().unwrap();
+
+        if state == ConnectionState::Activated {
+            let state = connection.deactivate().unwrap();
+            assert_eq!(ConnectionState::Deactivated, state);
+
+            ::std::thread::sleep(::std::time::Duration::from_secs(5));
+
+            let state = connection.activate().unwrap();
+            assert_eq!(ConnectionState::Activated, state);
+
+            ::std::thread::sleep(::std::time::Duration::from_secs(5));
+        } else {
+            let state = connection.activate().unwrap();
+            assert_eq!(ConnectionState::Activated, state);
+
+            ::std::thread::sleep(::std::time::Duration::from_secs(5));
+
+            let state = connection.deactivate().unwrap();
+            assert_eq!(ConnectionState::Deactivated, state);
+
+            ::std::thread::sleep(::std::time::Duration::from_secs(5));
```

Comment:
The NetworkManager service goes sometimes into unstable state after those calls and I added those sleeps temporary for our tests to pass, but we definitely need a better solution to handle this issue.